src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -1,31 +1,124 @@
+          let diff =
+            External_transition.staged_ledger_diff external_transition
+          in
+          let%map _, _, `Updated_staged_ledger staged_ledger =
+            Staged_ledger.apply ~logger staged_ledger diff
+          in
+          let new_breadcrumb =
+            Transition_frontier.Breadcrumb.create transition_with_hash
+              staged_ledger
+          in
+          (staged_ledger, new_breadcrumb :: acc) )
+    in
+    List.rev breadcrumbs
+
+  let materilize_breadcrumbs ~frontier ~logger = function
+    | [] -> `Success []
+    | root_transition :: _ as external_transitions -> (
+        let initial_state_hash =
+          With_hash.data root_transition
+          |> External_transition.protocol_state
+          |> External_transition.Protocol_state.previous_state_hash
+        in
+        match Transition_frontier.find frontier initial_state_hash with
+        | None -> `No_matching_root_hash initial_state_hash
+        | Some initial_breadcrumb -> (
+            let initial_staged_ledger =
+              Transition_frontier.Breadcrumb.staged_ledger initial_breadcrumb
+            in
+            match
+              construct_breadcrumb_path ~logger initial_staged_ledger
+                external_transitions
+            with
+            | Ok result -> `Success result
+            | Error e -> `Error e ) )
+
+  let choose_first_success (list : 'a option Deferred.t list) :
+      'a option Deferred.t =
+    let first_success = Ivar.create () in
+    List.iter list ~f:(fun x ->
+        Deferred.upon x (Option.iter ~f:(Ivar.fill first_success)) ) ;
+    Deferred.any
+      [ Ivar.read first_success >>| Option.some
+      ; Deferred.all list >>| Fn.const None ]
+
+  let get_transitions_from_peers ~logger ~peers hash =
+    List.map peers ~f:(fun peer ->
+        match%map Catchup_worker.dispatch Worker.Worker_rpc.rpc hash peer with
+        | Ok result -> result
+        | Error e ->
+            Logger.info logger
+              !"Connection error with peer %{sexp: Host_and_port.t}: %s"
+              peer (Error.to_string_hum e) ;
+            None )
+
+  let run ~logger ~network ~frontier ~catchup_job_reader
+      ~catchup_breadcrumbs_writer =
+    Strict_pipe.Reader.iter catchup_job_reader ~f:(fun transition_with_hash ->
+        let hash = With_hash.hash transition_with_hash in
+        (* TODO: pass in peers as paramet
+        er *)
+        let peers =
+          Network.peers network |> List.map ~f:Kademlia.Peer.external_rpc
+        in
+        let breadcrumbs =
+          List.map
+            (List.zip_exn peers
+               (get_transitions_from_peers ~logger ~peers hash))
+            ~f:(fun (peer, transitions_deferred) ->
+              let open Deferred.Option.Let_syntax in
+              let%bind transitions = transitions_deferred in
+              Deferred.return
+              @@
+              match materilize_breadcrumbs ~frontier ~logger transitions with
```

Comment:
This is fine for the naive one, but we should make an issue to fix this logic later. Right now, we ask all peers for all the transitions back to their root, and materialize breadcrumbs for all of the responses. But this is very wasteful, both from a networking and CPU perspective (materializing breadcrumbs is not a cheap operation). With the new networking work, there is a concept of an `Envelope` which we can use to access the sender of a message, and we can write this logic to start off by just asking the sender instead of all. If we still need to fan out and ask multiple, we should probably stop generating breadcrumbs for responses once we find a good response. You have some of that logic here, but I'm fairly certain that `choose_first_success` will not interrupt the other `Deferred.t` values that were not selected.